File: sorting_searching_and_logs/find_a_duplicate/find_a_duplicate.py
def find_repeat(numbers):
  # Since we must make sure space complexity is not abused, at best we can sort the array
  # and then find repeats that way with O(nlgn) complexity
  # but that destroys the input's order so a better way is to 
  # use binary search to search sub-ranges of 1 to n given that
  # the question says we have values distinct from 1 to n and a total
  # of n + 1 items
  # meaning there has to be a duplicate 
  # so binary search in half ranges, checking n items in the array for 
  # counts that belong in each range and if there are more counts for that range
  # than there are legal distinct numbers possible like 1...3
  # i.e. the count is 6 (n + 1) but the range is 1 to 5 and 6 > 5 
  floor = 1
  ceiling = len(numbers) - 1

  while floor < ceiling:
    midpoint = floor + (ceiling - floor) / 2
    lower_floor = floor 
    lower_ceiling = midpoint
    higher_floor = midpoint + 1
    higher_ceiling = ceiling

    # in original range it should be n distinct items
    # take the ceiling - floor + 1
    distinct_number_of_items_in_lower_range = lower_ceiling - lower_floor + 1

    number_of_items_in_lower_range = 0
    for number in numbers:
      if number >= lower_floor and number <= lower_ceiling:
        number_of_items_in_lower_range += 1
    
    if number_of_items_in_lower_range > distinct_number_of_items_in_lower_range:
      # this is the range we focus on
      floor = lower_floor
      ceiling = lower_ceiling 
    else:
      floor = higher_floor
      ceiling = higher_ceiling
  
  return floor

File: sorting_searching_and_logs/find_a_duplicate/find_a_duplicate.js
function findRepeat(numbers) {

  // Find a number that appears more than once
  // if I sort the array first (O(nlgn)), then I can just iterate thru the array one more time to get the repeat
  // space complexity: O(1), time complexity: O(nlgn) since its O(n) to find the dup but the sort takes nlgn

  numbers.sort();
  for (let i = 0; i < numbers.length - 1; i++) {
    let currNumber = numbers[i];
    let nextNumber = numbers[i+1];
    if (currNumber == nextNumber) {
      return currNumber;
    }
  }

}

function findRepeatKeepInputSame(numbers) {
  // Using needle in haystack
  // in the problem there are distinct numbers from 1 to n, and the total number of items is n + 1
  // that means if we split the array into subarrays of range (1...n/2) and (n/2 + 1...n)
  // sum of the distinct numbers for the subarray ranges will be n 
  // sum of the numbers count in the subarrays will be n + 1
  // it is understood that within any range, the count of numbers in that range will be n+1 while the 
  // number of distinct numbers will be n so there will be 1 or more duplicates in that range
  // using binary search, I can restrict the range and check the number of items in that range and if I
  // can find more numbers than the biggest number in that range, then I use use that floor and ceiling
  // until I compare a floor and ceiling of size 1, in which case I've found the number

  // smallest number is floor, biggest number is ceiling aka n
  let floor = 1;
  let ceiling = numbers.length - 1;

  while (floor < ceiling) {
    let midpoint = Math.floor( floor + (ceiling - floor) / 2);
    let lowerFloor = floor;
    let lowerCeiling = midpoint;
    let higherFloor = midpoint + 1;
    let higherCeiling = ceiling;

    const distinctNumberOfItemsInLowerRange = lowerCeiling - lowerFloor + 1;

    let numberOfItemsInLowerRange = 0;
    numbers.forEach((num) => {
      if (num >= lowerFloor && num <= lowerCeiling) {
        numberOfItemsInLowerRange += 1;
      }
    });


    if (numberOfItemsInLowerRange > distinctNumberOfItemsInLowerRange) {
      floor = lowerFloor;
      ceiling = lowerCeiling;
    } else {
      floor = higherFloor;
      ceiling = higherCeiling;
    }
  }

  return floor;
}